packages/aws-cdk/lib/api/util/cloudformation.ts

```diff
@@ -404,17 +413,49 @@ export class StackParameters {
     // Just append all supplied overrides that aren't really expected (this
     // will fail CFN but maybe people made typos that they want to be notified
     // of)
-    const unknownParam = ([key, _]: [string, any]) => this.params[key] === undefined;
+    const unknownParam = ([key, _]: [string, any]) => this.formalParams[key] === undefined;
     const hasValue = ([_, value]: [string, any]) => !!value;
     for (const [key, value] of Object.entries(updates).filter(unknownParam).filter(hasValue)) {
+      this.values[key] = value!;
       this.apiParameters.push({ ParameterKey: key, ParameterValue: value });
     }
   }
 
   /**
    * Whether this set of parameter updates will change the actual stack values
    */
-  public get changed() {
+  public hasChanges(currentValues: Record<string, string>): boolean {
+    // If any of the parameters are SSM parameters, deploying must always happen
+    // because we can't predict what the values will be.
+    if (Object.values(this.formalParams).some(p => p.Type.startsWith('AWS::SSM::Parameter::'))) {
+      return true;
+    }
+
+    // Otherwise we're dirty if:
+    // - any of the existing values are removed, or changed
+    if (Object.entries(currentValues).some(([key, value]) => !(key in this.values) || value !== this.values[key])) {
+      return true;
+    }
+
+    // - any of the values we're setting are new
+    if (Object.keys(this.values).some(key => !(key in currentValues))) {
+      return true;
+    }
+
+    return false;
+  }
+}
+
+/**
+ * The values that change between the current and a new deployment
+ */
+export class ParameterChanges {
```

Comment:
I don't see this being referenced anywhere. Vestigial remains?

Reply:
GD DMN

yes